RFC-0010-dyllib-cpp.md

```diff
@@ -0,0 +1,131 @@
+
+RFC-0003 added the ability to validate C++ functions in the Presto coordinator without creating a corresponding function in the Java SPI. RFC 0005 allows users to register custom functions dynamically without having to fork prestissimo.
+
+### [Optional] Goals
+Register custom functions, types, and connectors dynamically without requiring a fork of Prestissimo.
+### [Optional] Non-goals
+Security concerns: There are some security concerns associated with using the built in plugins. Using the dlopen library, we run the risk of opening unsecure unknown shared objects especially given the lack of any form of validation. On C++, we share the same limitations as the functionality in Java with a noted exception of Java Presto running in a VM environmnet while the C++ version will be run locally. 
+
+Performance concerns: Users may write poorly-written custom elements which may degrade performance.
+
+Reliability concerns: Users may write poorly-written custom elements which may cause instability to the Presto cluster.
+
+This is a well-known limitation of unfenced functions. Users are aware of these risks which are also concerns of the existing Plugin SPI. The goal of this RFC is to bring feature parity with the Plugin SPI.
+## Proposed Implementation
+This library will be implemented in three stages, with the first one being for User Defined Functions. Type and Connectors will build onto the dynamic library following the same format in terms of the implementation and design.
+
+### General Implementation
+The user can register their custom elements dynamically by creating .dylib or .so shared libraries and dropping them in a plugin directory. This plugin directory needs to be defined with the plugin.dir property in config.properties of the prestissimo worker. This directory will be scanned and all shared libraries will be attempted to be dynamically loaded when the worker or the sidecar process starts.
+
+#### User Defined Function Implementation
+For dynamically loaded function registration, the format followed is mirrored of that of built-in function registration with some noted differences. For instance, the below example function uses the extern "C" keyword to protect against name mangling. Additionally, a registry() function call is also necessary here.
+
+```
+#include "presto_cpp/main/dynamic_registry/DynamicFunctionRegistrar.h"
+
+namespace facebook::presto::common::dynamicRegistry {
+
+template <typename T>
+struct DynamicFunction {
+  FOLLY_ALWAYS_INLINE bool call(int64_t& result) {
+    result = 123;
+    return true;
+  }
+};
+
+} // namespace facebook::presto::common::dynamicRegistry
+
+extern "C" {
+void registry12() {
+  facebook::presto::registerPrestoFunction<
+      facebook::presto::common::dynamicRegistry::DynamicFunction,
+      int64_t>("dynamic");
+}
+}
+```
+To turn these files into shared libraries, the CMakeLists.txt file will be used.
+
+#### Error and Signal Handling
```

Comment:
I actually think we need to change this behavior.  For example, if one worker crashes while loading the shared library while another succeeds, this would leave the cluster in an inconsistent state and would fail queries.  I think it's better fail loudly.